drs_cli/client.py

```diff
@@ -1,229 +1,424 @@
+from drs_cli.errors import (
+    InvalidObjectData, exception_handler,
+    InvalidObjectIdentifier,
+    InvalidResponseError,
+    InvalidURI,
+)
 
 logger = logging.getLogger(__name__)
+sys.excepthook = exception_handler
 
 
 class DRSClient():
-    """Client to communicate with a GA4GH DRS instance.
-    Supports additional endpoints defined in DRS-filer.
-    (https://github.com/elixir-cloud-aai/drs-filer)
+    """Client to communicate with a GA4GH DRS instance. Supports additional
+    endpoints defined in DRS-filer
+    (https://github.com/elixir-cloud-aai/drs-filer).
 
-    Args:
-        host: Base URL of DRS instance.
-        port: Port at which DRS instance can be accessed.
-        base-path: Path at which DRS endpoints can be accessed.
-        token: Bearer token for authentication.
+    Arguments:
+        uri: Either the base URI of the DRS instance to connect to in either
+            'https' or 'http' schema (note that fully compliant DRS instances
+            will use 'https' exclusively), e.g., `https://my-drs.app`, OR a
+            hostname-based DRS URI, cf.
+            https://ga4gh.github.io/data-repository-service-schemas/preview/develop/docs/#_hostname_based_drs_uris.
+        port: Override default port at which the DRS instance can be accessed.
+            Only required for DRS instances that are not fully spec-compliant,
+            as the default port is defined in the DRS documentation, cf.
+            https://ga4gh.github.io/data-repository-service-schemas/preview/develop/docs/#_hostname_based_drs_uris.
+        base-path: Override default path at which the DRS API is accessible at
+            the given DRS instance. Only required for DRS instances that are
+            not fully spec-compliant, as the default port is defined in the DRS
+            documentation, cf.
+            https://ga4gh.github.io/data-repository-service-schemas/preview/develop/docs/#_hostname_based_drs_uris.
+        use_http: Set the URI schema of the DRS instance to `http` instead of
+            `https`when a DRS URI was provided to `uri`.
+        token: Bearer token to send along with DRS API requests. Set if
+            required by DRS implementation. Alternatively, specify in API
+            endpoint access methods.
 
     Attributes:
-        url: URL to DRS endpoints, composed of `host`, `port` and `base_path`,
-            e.g.,"https://my-drs.app:8080/ga4gh/drs/v1"
-        token: Bearer token for authentication.
+        uri: URI to DRS endpoints, built from `uri`, `port` and `base_path`,
+            e.g.,"https://my-drs.app:443/ga4gh/drs/v1".
+        token: Bearer token for gaining access to DRS endpoints.
         headers: Dictionary of request headers.
     """
+    # set regular expressions as private class variables
+    _RE_DOMAIN_PART = r'[a-z0-9]([a-z0-9-]{1,61}[a-z0-9])?'
+    _RE_DOMAIN = rf"({_RE_DOMAIN_PART}\.)+{_RE_DOMAIN_PART}\.?"
+    _RE_DRS_ID = r'.+'
```

Comment:
Yes, correct. But if you read on, it says:
> DRS IDs can contain other characters, but they MUST be encoded into valid DRS IDs whenever they are used in API calls. This is because non-encoded IDs may interfere with the interpretation of the objects/{id}/access endpoint. To overcome this limitation use percent-encoding of the ID, see RFC 3986 § 2.4

This is why I am not imposing any rules for characters after the slash following the hostname (`drs://host.name/<#!HERE_BE_DRAGONS#!>`). Instead, I have added some lines to percent-encode the ID as suggested (the `quote(string='xyz', safe='')` stuff).